pkg/tempo/automaticloggingprocessor/automaticloggingprocessor.go

```diff
@@ -0,0 +1,242 @@
+
+	typeSpan    = "span"
+	typeRoot    = "root"
+	typeProcess = "process"
+)
+
+type automaticLoggingProcessor struct {
+	nextConsumer consumer.TracesConsumer
+	cfg          *AutomaticLoggingConfig
+	lokiChan     chan<- api.Entry
+	done         atomic.Bool
+
+	logger log.Logger
+}
+
+func newTraceProcessor(nextConsumer consumer.TracesConsumer, cfg *AutomaticLoggingConfig) (component.TracesProcessor, error) {
+	logger := log.With(util.Logger, "component", "tempo automatic logging")
+
+	if nextConsumer == nil {
+		return nil, componenterror.ErrNilNextConsumer
+	}
+
+	if !cfg.Roots && !cfg.Processes && !cfg.Spans {
+		return nil, errors.New("automaticLoggingProcessor requires one of roots, processes, or spans to be enabled")
+	}
+
+	cfg.Overrides.LokiTag = override(cfg.Overrides.LokiTag, defaultLokiTag)
+	cfg.Overrides.ServiceKey = override(cfg.Overrides.ServiceKey, defaultServiceKey)
+	cfg.Overrides.SpanNameKey = override(cfg.Overrides.SpanNameKey, defaultSpanNameKey)
+	cfg.Overrides.StatusKey = override(cfg.Overrides.StatusKey, defaultStatusKey)
+	cfg.Overrides.DurationKey = override(cfg.Overrides.DurationKey, defaultDurationKey)
+	cfg.Overrides.TraceIDKey = override(cfg.Overrides.TraceIDKey, defaultTraceIDKey)
+
+	return &automaticLoggingProcessor{
+		nextConsumer: nextConsumer,
+		cfg:          cfg,
+		logger:       logger,
+		done:         atomic.Bool{},
+	}, nil
+}
+
+func (p *automaticLoggingProcessor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	rsLen := td.ResourceSpans().Len()
+	for i := 0; i < rsLen; i++ {
+		rs := td.ResourceSpans().At(i)
+		ilsLen := rs.InstrumentationLibrarySpans().Len()
+
+		var svc string
+		svcAtt, ok := rs.Resource().Attributes().Get(conventions.AttributeServiceName)
+		if ok {
+			svc = svcAtt.StringVal()
+		}
+
+		for j := 0; j < ilsLen; j++ {
+			ils := rs.InstrumentationLibrarySpans().At(j)
+			spanLen := ils.Spans().Len()
+
+			lastTraceID := ""
+			for k := 0; k < spanLen; k++ {
+				span := ils.Spans().At(k)
+				traceID := span.TraceID().HexString()
+
+				if p.cfg.Spans {
+					p.exportToLoki(typeSpan, traceID, p.spanKeyVals(span, svc)...)
+				}
+
+				if p.cfg.Roots && span.ParentSpanID().IsEmpty() {
+					p.exportToLoki(typeRoot, traceID, p.spanKeyVals(span, svc)...)
+				}
+
+				if p.cfg.Processes && lastTraceID != traceID {
+					lastTraceID = traceID
+					p.exportToLoki(typeProcess, traceID, p.processKeyVals(rs.Resource(), svc)...)
+				}
+			}
+		}
+	}
+
+	return p.nextConsumer.ConsumeTraces(ctx, td)
+}
+
+func (p *automaticLoggingProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{}
+}
+
+// Start is invoked during service startup.
+func (p *automaticLoggingProcessor) Start(ctx context.Context, _ component.Host) error {
+	loki := ctx.Value(contextkeys.Loki).(*loki.Loki)
+	if loki == nil {
+		return fmt.Errorf("key %s does not contain a Loki instance", contextkeys.Loki)
+	}
+	lokiInstance := loki.Instance(p.cfg.LokiName)
```

Comment:
What happens if the Agent config gets changed and the LokiName used here goes away? We'll need some kind of validation somewhere at the pkg/config level to make sure you can't break this. 

It *may* be easier in the short-term to launch our own promtail from within here instead of referencing one from the logging block, but it's up to you.

Reply:
In that case ApplyConfig would fail: 
https://github.com/grafana/agent/blob/main/pkg/tempo/tempo.go#L48

which means that this would fail as well?
https://github.com/grafana/agent/blob/main/cmd/agent/entrypoint.go#L101

I think that's ok?

Reply:
So, the main problem here is if a user goes through these steps:

1. User starts the Agent using the automatic logger 
2. User modifies Agent config, removes the Loki entry, but the Tempo config is left unchanged 
3. User reloads the Agent. 
4. Automatic logging processor breaks, referencing a Promtail that has shut down

This is because we [cache the input to ApplyConfig](https://github.com/grafana/agent/blob/e874ac465849e33ddb77bd121ce5357bdf894b99/pkg/tempo/instance.go#L54) so we're not always restarting systems that didn't change.

Reply:
(But if we have config validation in the [global ApplyDefaults](https://github.com/grafana/agent/blob/e874ac465849e33ddb77bd121ce5357bdf894b99/pkg/config/config.go#L51), we can return an error before any part of the system starts to reload, which would leave it in a working state)